include/deal.II/base/discrete_time.h

```diff
@@ -78,24 +87,178 @@ DEAL_II_NAMESPACE_OPEN
+ *   $x$ depends on the calculation of another variable such as $y$, then $y$
+ *   must be updated before $x$ can be updated. The question arises whether
+ *   time should be incremented before updating state quantities. Multiple
+ *   possibilities exist, depending on program and formulation requirements,
+ *   and possibly the programmer's preferences:
+ *   * Time is incremented before the rest of the updates. In this case, even
+ *     though time is incremented to $t_{n+1}$, not all variables are updated
+ *     yet. During this update phase, $dt$ equals the *previous* time step
+ *     size. *Previous* means that it is referring to the $dt$ of the
+ *     `advance_time()` command that was performed previously. In the
+ *     following example code, we are assuming that `a` and `b` are two state
+ *     variables that need to be updated in this time step.
+ *     @code
+ *       time.advance_time();
+ *       new_a = update_a(a, b, time.get_previous_step_size());
+ *       b = update_b(a, b, time.get_previous_step_size());
+ *       a = new_a;
+ *     @endcode
+ *   * Time is incremented from $t_n$ to $t_{n+1}$ after all variables have
+ *     already been updated for $t_{n+1}$. During the update stage, $dt$ is
+ *     denoted as the *next* time step size. *Next* means that $dt$ of the
+ *     step corresponds to the `advance_time()` command that will happen
+ *     subsequently.
+ *     @code
+ *       new_a = update_a(a, b, time.get_next_step_size());
+ *       b = update_b(a, b, time.get_next_step_size());
+ *       a = new_a;
+ *       time.advance_time();
+ *     @endcode
+ *   * Time is incremented in the middle of the other updates: In this case
+ *     $dt$ would correspond to *next* or *previous* depending of whether it
+ *     is used before or after the call to `advance_time()`.
+ *     @code
+ *       new_a = update_a(a, b, time.get_next_step_size());
+ *       time.advance_time();
+ *       b = update_b(a, b, time.get_previous_step_size());
+ *       a = new_a;
+ *     @endcode
+ *
+ * One thing to note is that, during the update phase, $dt$ is referred to
+ * either **next** or **previous** time step size, depending on whether the
+ * command `advance_time()` has been called yet. The notion of *current* time
+ * step size is ill-defined. In fact, in the update stage the definition of
+ * every variable depends on whether it has been updated yet or not, hence the
+ * name **the inconsistent stage**.
+ *
+ * The following code snippet shows the code sections for the snapshot stage
+ * and the update stage in the context of a complete time-dependent
+ * simulation. This code follows the coding conventions incorporated in the
+ * tutorial examples.
+ * @code
+ * // pre-processing/setup stage {
+ * make_grid();
+ * setup_system();
+ * for (DiscreteTime time(0., 1., 0.1);  // } end pre-processing/setup stage
+ *      time.is_at_end() == false;
+ *      time.advance_time())             // part of the update stage
```

Comment:
```suggestion
 *      time.advance_time())             // part of the update stage; happens at the end of the loop body's execution
```

Reply:
Actually, I'm not too fond of writing the loop like this. I'd much rather the `time.advance_time()` call was explicitly put at the end of the loop body, and the initialization of the `time` variable moved before the `for` loop. The `for` loop would then simply become a `while` loop. I think that would be easier to read.

Reply:
Hmm... maybe you and @rezarastak need to negotiate on this. It seems to have been the purpose of the whole discussion, and is also aligned with the existing documentation in the class intro.

Reply:
@bangerth What you've stated here is what I've always done timestepping in the past, but I'm trying to come around to @rezarastak's way of doing things. Clearly there's merit and a reason for using any of the permissible permutations of constructing a time-stepping loop. I like that the `for` loop is structured and self-contained (i.e. drives its own incrementation). In its simplest form it encapsulates, without the possibility for misinterpretation, the full set of operations required from the start of _simulation_ to the end of _simulation_.

Reply:
@bangerth @rezarastak Would you feel it to be of value if we write the equivalent loop using a `while loop` as an additional example? If so then I can add it as a follow up to this PR.

Reply:
I am happy with the current `for loop` style and I think users who read it can follow the order of operations happening within the loop. Using `for loop` is good for consistency with the beginning of the tutorial as @jppelteret said. Regardless, the users of this class can rewrite it in while loop style. The style preference is very subjective.

I can add to the documentation that this loop can also be written in `while` format and `do { ... } while` format as shown in step-21.

Reply:
Leave it as you had it and I can follow up with a patch that shows an alternative. I should give you some artistic license :-)